README.md
Gearman
===================

[![CPAN version](https://badge.fury.io/pl/Gearman.png)](https://badge.fury.io/pl/Gearman)
[![Build Status](https://travis-ci.org/p-alik/perl-Gearman.png)](https://travis-ci.org/p-alik/perl-Gearman)
[![Coverage Status](https://coveralls.io/repos/github/p-alik/perl-Gearman/badge.png)](https://coveralls.io/github/p-alik/perl-Gearman)

This repository contains perl [Gearman](http://gearman.org) Client/Worker implementation

perl Modules
------------
* [Gearman::Client](https://metacpan.org/pod/Gearman::Client) - Client for gearman distributed job system
* [Gearman::Task](https://metacpan.org/pod/Gearman::Task)     - a task in Gearman, from the point of view of a client
* [Gearman::Worker](https://metacpan.org/pod/Gearman::Worker) - Worker for gearman distributed job system



CHANGES
2.004.009 2017-10-06 16:01:00 Europe/Berlin
    - bug fixing Worker->work: register functions on reconnect.
      see: https://github.com/p-alik/perl-Gearman/issues/23
    - Gearman::Client->do_task invokes Gearman::Task on_complete hook, if it is defined
    - add documentation of Gearman::Client exceptions property

2.004.008 2017-06-13 13:35:00 Europe/Berlin
    - bug fixing: double worker registration
      https://github.com/p-alik/perl-Gearman/issues/21
    - tests start gearmand with --log-file option. Default /dev/stderr

2.004.007 2017-05-30 10:51:00 Europe/Berlin
    - bug fix: stringify job server in
      - get_job_server_status
      - get_job_server_clients
      - get_job_server_jobs

2.004.006 2017-05-25 18:39:00 Europe/Berlin
    - bug fixing: Taskset->add_task support for large data
    - ssl tests large task data and large work result

2.004.005 2017-05-19 13:11:00 Europe/Berlin
    - new worker method send_work_exception
    - send_work_fail does not support message parameter

2.004.004 2017-05-09 07:53:00 Europe/Berlin
    - refactoring: Taskset carefully loops over loaned_sock

2.004.003 2017-04-11 22:24:00 Europe/Berlin
    - Client and Worker could set SSL ca_file parameter
    - Taskset.pm refactoring:
      replace fileno handling by IO::Select

2.004.002 2017-04-11 11:01:00 Europe/Berlin
    - do not force Client and Worker to set useless SSL ca_certs parameter

2.004.001 2017-04-11 14:38:00 Europe/Berlin
     - Gearman::Client and Gearman::Worker could set prefix separator
       - It would be recommended to replace default prefix separator to insure
         gearmand persistent queue recovery.

2.003.002 2017-03-26 22:01:00 Europe/Berlin
     - tests are perl 5.25 ready (use lib .)
     -- GEARMAN_WORKER_USE_STDIO test only with AUTHOR_TESTING

2.003.001 2017-03-13 21:47:00 Europe/Berlin
     -- test compatibility for gearmand and Gearman::Server

2.003_002 2017-03-02 14:45:00 Europe/Berlin
     -- restore Gearman::Client::Async support. #10
     -- try to bug fix ssl callback testing

2.003_001 2017-02-18 18:37:00 Europe/Berlin
     -- worker:
        -- remove THROW_EXCEPTIONS defination in BEGIN block
        -- add support of protocol commands WORK_DATA and WORK_WARNING
        -- new methods:
           -- send_work_data
           -- send_work_warning
           -- send_work_complete
           -- send_work_fail
        -- _get_js_sock: no execution of object _on_connect if callback is defined
     -- fixing client do_task: do not overwrite on_fail callback
     -- Gearman::Task support of low priority
     -- remove Ref::Util dependency

2.002.004 2016-12-05 09:02:16 Europe/Berlin
     -- requires Ref::Util >= 0.020; see cpan tester report 848fbc3e-b94f-11e6-8baf-a6caeafb6ca6
     -- Makefile.PL updates
        LICENSE: perl
        min perl version 5.8.1
     -- fixing worker GEARMAN_WORKER_USE_STDIO env test

2.002.003 2016-12-04 22:13:00 Europe/Berlin
     -- add DISTNAME into Makefile.PL to get back to the old distribution name

2.002.002 2016-12-03 12:10:00 Europe/Berlin
     -- skip worker stop_if test if no gearmand

2.002.001 2016-12-02 22:01:00 Europe/Berlin (TRIAL RELEASE)
     -- job_servers entries could be a hash references or string
     -- bug fixing Gearman::Util::send_req() SSL frame limitation (esabol)

2.001.001 2016-10-13 10:45:00 Europe/Berlin (TRIAL RELEASE)
     -- support for SSL connection to gearmand
     -- support IPv6
     -- tests refactoring

1.130.004 (2016-08-06)
     -- add Gearman::ResponseParser pod
     -- add Gearman::ResposeParser::Taskset pod
     -- rename Gearman::Taskset->_process_packet to process_packet
     -- issue 116744 Utility pod

1.130.003 (2016-08-05)
     -- check OS in Test::Gearman and don't use "which" on Windows
     -- replace 'use Errno qw(EAGAIN)' by 'POSIX qw(:errno_h)'
        see: http://www.cpantesters.org/cpan/report/d801a704-5975-11e6-9451-9b92aab8e0c0

1.130.002 (2016-08-03)
     -- skip worker _get_js_sock test without gearmand

1.130.001 (2016-08-02)
     -- refactoring of Gearman::Util::read_res_packet: no goto/redo
     -- test suite refactoring
     -- solved issue 85191 Programming error prevents Gearman::Worker::work() from connecting to servers
     -- solved issue 59185 document Gearman::Client::get_status
     -- solved issue 56508 using $task->fail("reason goes here") in Gearman::Taskset

1.12.009 (2016-06-04)
     -- run t/30-maxqueue.t and t/40-prefix.t only with AUTHOR_TESTING
     -- s/::Object/::Objects/

1.12.008 (2016-06-03)
     -- rename Gearman::Base back to Geamrman::Object because Gearman-Client-Async depends on it

1.12.007 (2016-05-20)
     -- more tests only with AUTHOR_TESTING

1.12.006 (2016-05-19)
     -- tests refactoring
     -- execute some tests only if AUTHOR_TESTING env is defined

1.12.005 (2016-05-06)
     -- rm META.*

1.12.004 (2016-05-06)

     -- solved issue 5324ce04-0bae-11e6-a317-839d20bbf307
        there was a bug in test script

1.12.003 (2016-05-01)

     -- tested with perl5-22-1 to improve cpan tester resport
        5324ce04-0bae-11e6-a317-839d20bbf307

     -- fields moved from Gearman::Objects to proper modules

     -- rm obsolete Gearman::Objects

     -- use warnings (except redefine)

     -- perltidy applied to all modules

1.12.002 (2014-12-19)

     -- attempt to fix bugs: 89037, 100594, 101012

     -- cpan tester report 586f5968-8489-11e4-adb9-802de0bfc7aa is - ugly - fixed too

     -- refactoring: Client/Worker base class added

     -- dynamic tests by using environment variable GEARMAN_SERVERS

1.12 (2014-12-14)

     -- Repoint HACKING file.

     -- Make a jobserver connection use the command timeout during exception negotiation.

     -- Make $taskset->add_task use the command timeout to not hang during job submission.

     -- Add a client option 'command_timeout' to indicate how long we should wait before considering a

Util.pm

package Gearman::Util;
use strict;

# I: to jobserver
# O: out of job server
# W: worker
# C: client of job server
# J : jobserver
our %cmd = (
            1 =>  [ 'I', "can_do" ],     # from W:  [FUNC]
            2 =>  [ 'I', "cant_do" ],    # from W:  [FUNC]
            3 =>  [ 'I', "reset_abilities" ],  # from W:  ---
            22 => [ 'I', "set_client_id" ],    # W->J: [RANDOM_STRING_NO_WHITESPACE]
            4 =>  [ 'I', "pre_sleep" ],  # from W: ---

            6 =>  [ 'O', "noop" ],        # J->W  ---
            7 =>  [ 'I', "submit_job" ],    # C->J  FUNC[0]UNIQ[0]ARGS
            21 =>  [ 'I', "submit_job_high" ],    # C->J  FUNC[0]UNIQ[0]ARGS
            18 => [ 'I', "submit_job_bg" ], # C->J     " "   "  " "

            8 =>  [ 'O', "job_created" ], # J->C HANDLE
            9 =>  [ 'I', "grab_job" ],    # W->J --
            10 => [ 'O', "no_job" ],      # J->W --
            11 => [ 'O', "job_assign" ],  # J->W HANDLE[0]FUNC[0]ARG

            12 => [ 'IO',  "work_status" ],   # W->J/C: HANDLE[0]NUMERATOR[0]DENOMINATOR
            13 => [ 'IO',  "work_complete" ], # W->J/C: HANDLE[0]RES
            14 => [ 'IO',  "work_fail" ],     # W->J/C: HANDLE

            15 => [ 'I',  "get_status" ],  # C->J: HANDLE
            20 => [ 'O',  "status_res" ],  # C->J: HANDLE[0]KNOWN[0]RUNNING[0]NUM[0]DENOM

            16 => [ 'I',  "echo_req" ],    # ?->J TEXT
            17 => [ 'O',  "echo_res" ],    # J->? TEXT

            19 => [ 'O',  "error" ],       # J->? ERRCODE[0]ERR_TEXT
            );

our %num;  # name -> num
while (my ($num, $ary) = each %cmd) {
    die if $num{$ary->[1]};
    $num{$ary->[1]} = $num;
}

sub cmd_name {
    my $num = shift;
    my $c = $cmd{$num};
    return $c ? $c->[1] : undef;
}

sub pack_req_command {
    my $type_arg = shift;
    my $type = $num{$type_arg} || $type_arg;
    die "Bogus type arg of '$type_arg'" unless $type;
    my $arg = $_[0] || '';
    my $len = length($arg);
    return "\0REQ" . pack("NN", $type, $len) . $arg;
}

sub pack_res_command {
    my $type_arg = shift;
    my $type = int($type_arg) || $num{$type_arg};
    die "Bogus type arg of '$type_arg'" unless $type;
    my $len = length($_[0]);
    return "\0RES" . pack("NN", $type, $len) . $_[0];
}

# returns undef on closed socket or malformed packet
sub read_res_packet {
    my $sock = shift;
    my $err_ref = shift;

    my $buf;
    my $rv;

    my $err = sub {
        my $code = shift;
        $$err_ref = $code if ref $err_ref;
        return undef;
    };

    return $err->("malformed_header") unless sysread($sock, $buf, 12) == 12;
    my ($magic, $type, $len) = unpack("a4NN", $buf);
    return $err->("malformed_magic") unless $magic eq "\0RES";

    if ($len) {
        $rv = sysread($sock, $buf, $len);
        return $err->("short_body") unless $rv == $len;
    }

    $type = $cmd{$type};
    return $err->("bogus_command") unless $type;
    return $err->("bogus_command_type") unless index($type->[0], "O") != -1;

    return {
        'type' => $type->[1],
        'len' => $len,
        'blobref' => \$buf,
    };
}

sub send_req {
    my ($sock, $reqref) = @_;
    return 0 unless $sock;

    my $len = length($$reqref);
    #TODO: catch SIGPIPE
    my $rv = $sock->syswrite($$reqref, $len);
    return 0 unless $rv == $len;
    return 1;
}

# given a file descriptor number and a timeout, wait for that descriptor to
# become readable; returns 0 or 1 on if it did or not
sub wait_for_readability {
    my ($fileno, $timeout) = @_;
    return 0 unless $fileno && $timeout;

    my $rin = '';
    vec($rin, $fileno, 1) = 1;
    my $nfound = select($rin, undef, undef, $timeout);

    # nfound can be undef or 0, both failures, or 1, a success
    return $nfound ? 1 : 0;
}

1;
